src/coreclr/jit/fgwasm.cpp

```diff
@@ -123,13 +167,805 @@ FlowGraphDfsTree* Compiler::fgWasmDfs()
+                }
+
+                netLikelihood += dispatchToHeaderEdge->getLikelihood();
+
+                succs[headerNumber] = dispatchToHeaderEdge;
+                cases[headerNumber] = dispatchToHeaderEdge;
+
+                headerNumber++;
+            }
+
+            // Create the dispatch switch... really there should be no default but for now we'll have one.
+            //
+            JITDUMP("Dispatch header is " FMT_BB "; %u cases\n", dispatcher->bbNum, numHeaders);
+            BBswtDesc* const swtDesc =
+                new (m_comp, CMK_BasicBlock) BBswtDesc(succs, numHeaders, cases, numHeaders, true);
+            dispatcher->SetSwitch(swtDesc);
+
+            GenTree* const   controlVar = m_comp->gtNewLclvNode(controlVarNum, TYP_INT);
+            GenTree* const   switchNode = m_comp->gtNewOperNode(GT_SWITCH, TYP_VOID, controlVar);
+            Statement* const switchStmt = m_comp->fgNewStmtAtEnd(dispatcher, switchNode);
+
+            m_comp->gtSetStmtInfo(switchStmt);
+            m_comp->fgSetStmtSeq(switchStmt);
+        }
+
+        // Handle nested Sccs
+        //
+        for (Scc* const nested : m_nested)
+        {
+            modified |= nested->TransformViaSwitchDispatch();
+        }
+
+        return modified;
+    }
+};
+
+//-----------------------------------------------------------------------------
+// WasmFindSccs: find strongly connected components in the flow graph
+//
+// Arguments:
+//   sccs [out] -- top level Sccs in the flow graph
+//
+// Returns:
+//   true if the flow graph was modified
+//
+void FgWasm::WasmFindSccs(ArrayStack<Scc*>& sccs)
+{
+    assert(m_dfsTree->IsForWasm());
+    BitVec allBlocks = BitVecOps::MakeFull(&m_traits);
+    WasmFindSccsCore(allBlocks, sccs, m_dfsTree->GetPostOrder(), m_dfsTree->GetPostOrderCount());
+    unsigned numIrreducible = 0;
+
+    if (sccs.Height() > 0)
+    {
+        JITDUMP("\n*** Sccs\n");
+
+        for (int i = 0; i < sccs.Height(); i++)
+        {
+            Scc* const scc = sccs.Bottom(i);
+            scc->DumpAll();
+
+            numIrreducible += scc->NumIrr();
+        }
+    }
+    else
+    {
+        JITDUMP("\n*** No Sccs\n");
+    }
+
+    if (numIrreducible > 0)
+    {
+        JITDUMP("\n*** %u total Irreducible!\n", numIrreducible);
+    }
+}
+
+//-----------------------------------------------------------------------------
+// WasmFindSccsCore: find strongly connected components in a subgraph
+//
+// Arguments:
+//   subset  - bv describing the subgraph
+//   sccs    - [out] collection of Sccs found in this subgraph
+//   postorder - array of BasicBlock* in postorder
+//   postorderCount - size of the array
+//
+// Notes:
+//   Uses Kosaraju's algorithm: we walk the reverse graph starting
+//   from each block in reverse postorder.
+//
+void FgWasm::WasmFindSccsCore(BitVec& subset, ArrayStack<Scc*>& sccs, BasicBlock** postorder, unsigned postorderCount)
+{
+    SccMap              map(Comp()->getAllocator(CMK_WasmSccTransform));
```

Comment:
Will this map be sparse, or could it just be a flat map indexed by the postorder indices, since that mapping is dense?

Reply:
Initially this will need to cover the entire method, so I think we can create a flat array for that and then reuse it for subsequent subset cases.